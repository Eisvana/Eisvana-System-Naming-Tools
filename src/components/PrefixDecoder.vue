<script setup lang="ts">
import { eisvanaRegionAmount, systemIndexLength } from '../variables/constants';
import { regionGlyphs } from '../variables/regions';
import { computed, watchEffect } from 'vue';

import { usePrefixDataStore } from '../stores/prefixData';
import { storeToRefs } from 'pinia';

const prefixDataStore = usePrefixDataStore();
const { prefix } = storeToRefs(prefixDataStore);

const prefixParts = computed(() => prefix.value.toLowerCase().replace('ev', '').split('-'));

const regionNumber = computed(() => parseInt(prefixParts.value[0]));
const systemIndex = computed(() => {
  const actualNumber = prefixParts.value[1]?.replace('68+1', '69');
  return parseInt(actualNumber, 16);
});

const isNotEisvana = computed(
  () =>
    prefix.value &&
    prefixParts.value.length === 2 &&
    (isNaN(regionNumber.value) ||
      regionNumber.value < 1 ||
      regionNumber.value > eisvanaRegionAmount ||
      systemIndex.value.toString(16).length > systemIndexLength)
);

const glyphs = computed(() => {
  const glyphComponents = {
    planet: 0,
    system: systemIndex.value.toString(16).padStart(systemIndexLength, '0'),
    region: regionGlyphs[regionNumber.value - 1],
  };

  return Object.values(glyphComponents).join('').toUpperCase();
});

function resetTagDecoder() {
  prefix.value = '';
}

watchEffect(() => (prefix.value = prefix.value.toUpperCase()));
</script>

<template>
  <h2>Prefix Decoder</h2>
  <div>
    <label for="prefix">Prefix:</label>
    <input
      :aria-invalid="isNotEisvana || undefined"
      class="prefix-input"
      id="prefix"
      type="text"
      maxlength="9"
      v-model="prefix"
    />
  </div>
  <div
    class="error-wrapper"
    v-show="isNotEisvana"
  >
    <p class="error">Invalid Prefix!</p>
  </div>
  <div
    v-show="!isNaN(regionNumber) && !isNaN(systemIndex) && !isNotEisvana"
    class="output-wrapper"
  >
    <p class="output-label">Glyphs:</p>
    <output class="glyphs">{{ glyphs }}</output>
  </div>
  <button
    v-show="prefix"
    class="reset-button secondary"
    type="reset"
    @click="resetTagDecoder"
  >
    Reset
  </button>
</template>

<style scoped lang="scss">
.prefix-input {
  width: auto;
}
</style>
