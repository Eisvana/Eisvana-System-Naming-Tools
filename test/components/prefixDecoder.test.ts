import PrefixDecoder from '@/components/PrefixDecoder.vue';
import { beforeEach, describe, expect, it } from 'vitest';
import { mount, type VueWrapper } from '@vue/test-utils';
import { createTestingPinia } from '@pinia/testing';

describe('PrefixDecoder', () => {
  let wrapper: VueWrapper;
  const pinia = createTestingPinia();

  beforeEach(() => {
    wrapper = mount(PrefixDecoder, {
      global: {
        plugins: [pinia],
      },
    });
  });

  it('should render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
