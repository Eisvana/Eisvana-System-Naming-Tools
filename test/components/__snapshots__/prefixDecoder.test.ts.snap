// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PrefixDecoder > should render correctly 1`] = `
"<h2 data-v-29f6294f="">Prefix Decoder</h2>
<div data-v-29f6294f=""><label data-v-29f6294f="" for="prefix">Prefix:</label><input data-v-29f6294f="" class="prefix-input" id="prefix" type="text" maxlength="9"></div>
<div data-v-29f6294f="" class="status-wrapper" style="display: none;">
  <p data-v-29f6294f="" class="status error">Invalid Prefix!</p>
</div>
<div data-v-29f6294f="" class="output-wrapper" style="display: none;">
  <p data-v-29f6294f="" class="output-label">Glyphs:</p><output data-v-29f6294f="" class="glyphs">0NAN</output>
</div>
<button data-v-29f6294f="" class="reset-button secondary" type="reset" style="display: none;"> Reset </button>"
`;
